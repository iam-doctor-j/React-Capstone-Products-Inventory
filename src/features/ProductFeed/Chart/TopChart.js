import React from 'react';
import Card from 'react-bootstrap/Card';
import { connect } from 'react-redux';
import { Bar } from 'react-chartjs-2';
import * as _ from 'lodash';
import { refresh } from '../../Authentication/AuthenticationSlice';
import { fetchProductsFromDb } from '../ProductFeedSlice';
import Form from 'react-bootstrap/Form'

export class TopChart extends React.Component {

    constructor() {
        super();
        this.state = {
            chartData: {},
        };
        this.number = 3;
    }

    componentDidMount() {
        if(this.props.products.length === 0){
            this.props.dispatch(fetchProductsFromDb());
        }
        this.chart(this.number);
    }

    componentDidUpdate(prevProps, prevState) {
        if(JSON.stringify(this.props.products) !== JSON.stringify(prevProps.products)) {
            this.chart(this.number)
        }
    }

    chart = (number) => {
        console.log(number)
        let productName = [];
        let views = [];

        let sortedList = _.orderBy(this.props.products, ['views'], ['desc']);

        let topList = sortedList.slice(0, number);
        
        topList.forEach(item => {
            productName.push(item.name.value);
            views.push(item.views);
        });

        console.log(productName);
        console.log(views);

        this.setState({
            chartData: {
                labels: productName,
                datasets: [{
                    label: "Views",
                    data: views,
                    backgroundColor: ["rgba(75, 192, 192, 0.6)"],
                    borderWidth: 4,
                    minBarLength: 10,
                }]
            }
        });
    }

    render() {
        return(
            <div className="container">
                <div className="row justify-content-center">
                    <div className="col-12">
                        <Card>
                            <Card.Body>
                                <Card.Title>
                                <div className="d-flex">
                                    <span className="heading-text">Top Chart</span>
                                    <a className="ml-auto" onClick={() => {this.props.dispatch(refresh()); this.props.history.goBack()}} style={{cursor: 'pointer', color: 'var(--secondary)'}}><i className="fas fa-arrow-circle-left"></i></a>
                                </div>
                                </Card.Title>
                                {this.props.products.length >= 3 ? <div className="row mx-4 mt-4">
                                    <div className="col-md-4 col-12 label-text">Select number of top products</div>
                                    <div className="col-md-8 col-12">
                                        <Form.Control as="select" custom defaultValue={3} onChange={(e) => this.chart(+e.target.value)}>
                                            {/* <option value={1}>1</option>
                                            <option value={2}>2</option> */}
                                            <option value={3}>3</option>
                                            {this.props.products.length >= 4 ? <option value={4}>4</option> : <></>}
                                            {this.props.products.length >= 5 ? <option value={5}>5</option> : <></>}
                                            {this.props.products.length >= 6 ? <option value={6}>6</option> : <></>}
                                            {this.props.products.length >= 7 ? <option value={7}>7</option> : <></>}
                                            {this.props.products.length >= 8 ? <option value={8}>8</option> : <></>}
                                            {this.props.products.length >= 9 ? <option value={9}>9</option> : <></>}
                                            {this.props.products.length >= 10 ? <option value={10}>10</option> : <></>}
                                        </Form.Control>
                                    </div>
                                </div> : <></>}
                                <Bar 
                                    data={this.state.chartData}
                                    options={{
                                        scales:  {
                                            yAxes: [{
                                                ticks: {
                                                    beginAtZero: true
                                                }
                                            }]
                                        }
                                    }}
                                    />
                            </Card.Body>
                        </Card>
                    </div>
                </div>
            </div>
        );
    }
}

export default connect(state => {
    console.log(state);
    return {
        products: state.productFeed.products
    }
})(TopChart);