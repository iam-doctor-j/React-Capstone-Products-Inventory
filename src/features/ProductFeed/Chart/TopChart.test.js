import React from 'react';
import { shallow } from 'enzyme';
import { TopChart } from './TopChart';

describe('Test TopChart with shallow rendering', () => {
    let wrapper;

    const prods = `{
        "products": [
        {
            "id": "63796e43-f23b-542e-99a7-11db520c7bb6",
            "views": 6,
            "name": {
              "value": "Laptop",
              "visible": true
            },
            "description": {
              "value": "A simple Laptop",
              "visible": true
            },
            "manufacturer": {
              "value": "Lenovo",
              "visible": true
            },
            "price": {
              "value": 64999,
              "visible": true
            },
            "quantity": {
              "value": 1,
              "visible": true
            }
          },
          {
            "id": "d8548a39-7a6a-5258-bf6c-92fde0dc21fa",
            "views": 3,
            "name": {
              "value": "Mobile Phone",
              "visible": true
            },
            "description": {
              "value": "Smartphone",
              "visible": true
            },
            "manufacturer": {
              "value": "Realme",
              "visible": true
            },
            "price": {
              "value": 15000,
              "visible": true
            },
            "quantity": {
              "value": 1,
              "visible": true
            }
          },
          {
            "id": "cc61beb5-3315-5a82-9b34-da193d64a55c",
            "views": 4,
            "name": {
              "value": "Computer",
              "visible": true
            },
            "description": {
              "value": "A desktop PC fully assembled with all components. Mouse and Keyboard included.",
              "visible": true
            },
            "manufacturer": {
              "value": "HP",
              "visible": true
            },
            "price": {
              "value": 35000,
              "visible": true
            },
            "quantity": {
              "value": 1,
              "visible": false
            }
          },
          {
            "id": "32d9fd8f-91dd-5668-b003-83020ea5439f",
            "views": 4,
            "name": {
              "value": "Radio",
              "visible": true
            },
            "description": {
              "value": "A old radio",
              "visible": true
            },
            "manufacturer": {
              "value": "Saregama",
              "visible": true
            },
            "price": {
              "value": 1200,
              "visible": true
            },
            "quantity": {
              "value": 10,
              "visible": true
            }
        }
        ]
    }`;

    const props = JSON.parse(prods);

    beforeEach(() => {
        wrapper = shallow(<TopChart {...props} />);
    });

    it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    afterEach(() => {
        wrapper.unmount();
    })
})