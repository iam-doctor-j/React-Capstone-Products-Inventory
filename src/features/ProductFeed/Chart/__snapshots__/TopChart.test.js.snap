// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TopChart with shallow rendering should render correctly 1`] = `
<div
  className="container"
>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-12"
    >
      <Card
        body={false}
      >
        <CardBody>
          <CardTitle>
            <div
              className="d-flex"
            >
              <span
                className="heading-text"
              >
                Top Chart
              </span>
              <a
                className="ml-auto"
                onClick={[Function]}
                style={
                  Object {
                    "color": "var(--secondary)",
                    "cursor": "pointer",
                  }
                }
              >
                <i
                  className="fas fa-arrow-circle-left"
                />
              </a>
            </div>
          </CardTitle>
          <div
            className="row mx-4 mt-4"
          >
            <div
              className="col-md-4 col-12 label-text"
            >
              Select number of top products
            </div>
            <div
              className="col-md-8 col-12"
            >
              <FormControl
                as="select"
                custom={true}
                defaultValue={3}
                onChange={[Function]}
              >
                <option
                  value={3}
                >
                  3
                </option>
                <option
                  value={4}
                >
                  4
                </option>
              </FormControl>
            </div>
          </div>
          <Bar
            data={
              Object {
                "datasets": Array [
                  Object {
                    "backgroundColor": Array [
                      "rgba(75, 192, 192, 0.6)",
                    ],
                    "borderWidth": 4,
                    "data": Array [
                      6,
                      4,
                      4,
                    ],
                    "label": "Views",
                    "minBarLength": 10,
                  },
                ],
                "labels": Array [
                  "Laptop",
                  "Computer",
                  "Radio",
                ],
              }
            }
            options={
              Object {
                "scales": Object {
                  "yAxes": Array [
                    Object {
                      "ticks": Object {
                        "beginAtZero": true,
                      },
                    },
                  ],
                },
              }
            }
          />
        </CardBody>
      </Card>
    </div>
  </div>
</div>
`;
